src/util/settings.h

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_SETTINGS_H
+#define BITCOIN_UTIL_SETTINGS_H
+
+#include <map>
+#include <string>
+#include <vector>
+
+class UniValue;
+
+namespace util {
+
+//! Settings value type (string/integer/boolean/null variant).
```

Comment:
> Also key-value or array as requirements ? I suppose any other variant type should be easy to use with basic types..

ddd

Reply:
These definitely aren't not requirements of the type right now. There is no syntax for these things. Maybe this would be useful in the future for something like per-wallet settings (different preferences, RPC authentication setting), and I guess this is a little easier now, but otherwise the PR is neutral on this.

Reply:
> I suppose any other variant type should be easy to use with basic types..

I think `boost::variant` would be an inferior choice, but just as an experiment I tried dropping it in here with some helper methods to make it work without changes to other code:

```c++
struct SettingsValue : public boost::variant<std::nullptr_t, bool, int64_t, std::string>
{
    using variant::variant;
    template<typename T> const T* ptr() const { return boost::get<T>(this); }
    bool isNull() const { return ptr<std::nullptr_t>(); }
    bool isBool() const { return ptr<bool>(); }
    bool isNum() const { return ptr<int64_t>(); }
    bool isFalse() const { return ptr<bool>() && !*ptr<bool>(); }
    bool isTrue() const { return ptr<bool>() && *ptr<bool>(); }
    const std::string& get_str() const { return *ptr<std::string>(); }
    const bool& get_bool() const { return *ptr<bool>(); }
    const int64_t& get_int64() const { return *ptr<int64_t>(); }
};
```

This was sufficient for everything except `settings_test.cpp` (which uses serialization), so it confirms `SettingsValue` only requires storing ints, bools, strings and nulls, and the choice of type isn't very important.

Reply:
> I just listed the required methods and added a github link, but I can add the example code inline if preferred.

Looks great to me as it is. Thanks!